### ISSUE TITLE
Compression popping values that have not been pushed.

### ISSUE THREAD

**Issue by benclifford** [kazu-yamamoto/dns#103]
This snippet:

```
  rs <- makeResolvSeed defaultResolvConf {
          resolvInfo = RCHostName "90.155.94.93"
         }
  r <- withResolver rs $ \resolver -> lookupRaw resolver "www.cqx.ltd.uk" RRSIG

  print r
```

generates the attached pcap, and then fails with:

```
dugnutt: ParseError {errorContexts = [], errorMessage = "Failed reading: getDomain: 338", errorPosition = 1:449 (448)}
```

I put in some tracing around the push/pop statements and it looks like that location is not being `push`ed.

```
dugnutt www.cqx.ltd.uk RRSIG 
BENC: d pushing pos 24
BENC: d pushing pos 20
BENC: d pushing pos 16
BENC: d pushing pos 12
BENC: c pushing pos 32
BENC: c pushing pos 170
BENC: c pushing pos 308
dugnutt: ParseError {errorContexts = [], errorMessage = "Failed reading: getDomain': failed to resolve compression pointer to offset 338 at position 446", errorPosition = 1:449 (448)}
```

That server 90.155.94.93 is under my control and running bind 9.8.4.

pcap of the interaction:

www.hawaga.org.uk/tmp/dns-rrsig.pcap

**Comment by benclifford** [comment]
It looks like its a compression pointer at the start of the name of an authority section record, pointing into the body of an RRSIG body (where the appropriate string occurs), rather than into something that has been already decoded as a domain name by the `dns` library.

**Comment by kazu-yamamoto** [comment]
The current `dns` library supports *backward* pointers only, not *forward* pointers.
Let me consider whether or not forward pointers can be supported.
Anyway, this is a very good test case.
Thanks.

**Comment by benclifford** [comment]
@kazu-yamamoto The pointer here *is* backwards, I think: it points to an offset that is smaller than the current position returned by `getPosition`.

But, the pointer points into some opaque data in an RRSIG record. `getDomain'` has never been called to parse/`push` that data because RRSIGs are opaque in the current implementation; and so I think the pointer is pointing deep into an `UnknownRData` bytestring.

In principle this could happen with any rrtype which could hold a domain name but which the `dns` library does not know about.

**Comment by kazu-yamamoto** [comment]
OK. I understand. I will implement `RRSIG` anyway.

**Comment by kazu-yamamoto** [comment]
Cc: @vdukhovni

A generic solution would be storing the original DNS input to `SPut` and parsing the original DNS input according to DNS pointers if necessary. Some days ago, I gave up this approach because `IO` uses `conduit` which is streaming oriented, ie, the original DNS input data cannot be saved.

I decided to throw away `conduit` and now the original DNS input data can be saved. My PoC works well so far.
